### ISSUE TITLE
Custom material shaders

### ISSUE THREAD

**Issue by bdero** [bdero/flutter_scene#22]
Implement a workflow for users to write shaders that customize managed lighting models. To start off support `standard` and `unlit` lighting models.

## Why managed?
We could potentially go the route of making custom materials extremely simple by having the user straight up write the final Flutter GPU vertex/fragment shaders in their entirety. But such an approach would present an enormous usability tax on users:
- In order to keep the animation system working, users would be forced to add a bunch of confusing boilerplate functions or macro calls in their shaders (or just not support customizing the vertex stage, which IMO would be an unacceptable limitation).
- The vast majority of the time, users will want to customize an existing lighting model rather than entirely roll their own (but we'd still allow users to basically do this by providing an `unlit` lighting model). We can optimize for the common case by allowing the user to customize existing lighting models instead.

## Design proposal
- Users can import a `MaterialLibrary` using a Native Assets buildhook `buildMaterialLibrary`.
- The descriptor handed to `buildMaterialLibrary` is a list of paths to `.material` files.
- `buildMaterialLibrary` renders out a shader set for each material, with the vertex `vertex` stage containing at least two variations (for `skinned` and `unskinned`).
- `.material` files contain:
	- A `lighting_model` metadata field to specify the lighting model. For starters, support `standard` (default, uses Flutter Scene's PBR lighting) and `unlit` (a blank slate). This determines the writable globals in the fragment stage. All globals are optional and have reasonable defaults. (The globals below loosely resemble's Godot's convention for specifying in/out parameters)
	- Writable fragment globals are as follows:
		- `unlit` exposes `out vec4 COLOR` global.
		- `standard` exposes `out vec3 ALBEDO`, `out float ALPHA`, `out float METALLIC`, `out float ROUGHNESS`, `vec3 EMISSION`, `float AMBIENT_OCCLUSION`, and `vec3 NORMAL_MAP` (where `NORMAL_MAP` is the tangent space normal, as opposed to the vertex normal -- similar to other varying attributes, vertex normals are accessible via `in vec3 v_normal`).
	- For vertex shaders, the basic globals include: `inout vec3 MODEL_POSITION`, `inout vec3 NORMAL` (note: no tangents are necessary, we currently work out the tangent space while rendering), and `inout vec2 TEXTURE_COORDS`. All of these globals are already set to the input geometry data before `void vertex()` (the user's code) is called. We can consider adding flags later for e.g. setting the world position directly instead.
	- Metadata lists specifying custom uniform inputs (including textures), and which stages each uniform intends to target) as well as shared varying parameters. Uniform block fields can have defaults.
	- A shader section, containing GLSL code.
		- The shader section can optionally contain `void vertex()` and/or `void fragment()`.
		- The vertex/fragment stages have documented sets of globals available (as specified above).
- At runtime, material libraries can be loaded with `/* static */ MaterialLibrary.fromAsset('mylibrary.matlib')`.
- Materials of the correct base type can be instantiated with `matlib.createMaterial('custommat')`.
- Internally, the `Material`  is handed the vert/frag gpu.Shaders along with a custom uniform map.
- `Material` provides functionality for setting custom uniform parameters (which are type checked in debug builds). E.g. `material.setUniform(String name, dynamic value)`. Uniform layout and allocation management are internal concerns.

**Comment by bdero** [comment]
**Status update (revised).** v1 has shipped and custom materials work today. This comment is now the living design for v2 and v3, revised for two things we have learned since it was first written: what the simple raw-shader path actually feels like in practice, and the upstream Flutter GPU shader hot reload work, which is now in flight and is producer-agnostic. That second point materially changes v3 (it mostly evaporates into upstream work) and de-risks v2. The headline thesis is unchanged.

The headline thesis: flutter_scene should build a small declarative material format with a Dart-side preprocessor that emits standard GLSL for `impellerc`. We should not avoid the preprocessor for the sake of expedience, because the preprocessor is the only mechanism that gets us the authoring workflow we actually want. The alternative ("write a raw fragment shader, bind every uniform manually, redeclare every engine input by hand") is what Three.js's `ShaderMaterial` does, and the community feedback there is universally that it's painful enough to chase users away from custom shaders entirely. What we want, ergonomically, is closer to Godot's `.gdshader` or Filament's `.mat`: declare parameters once, fill in a small `surface()` function, get a fully wired-up material with name-based property bindings and inspector-friendly metadata.

## What's worth borrowing from prior art

**Filament's `.mat` file format** is the strongest published prior art for "declarative material with embedded shader code." A `.mat` file has three top-level blocks: a `material { }` metadata block (parameters, render state, shading model, vertex attribute requirements), an optional `vertex { }` GLSL block, and a `fragment { }` GLSL block whose entry point is a single `void material(inout MaterialInputs material)` function. Picking `shadingModel : lit` obligates the user to populate the `lit` subset of `MaterialInputs` (baseColor, metallic, roughness, normal, emissive, etc.); the engine takes care of the BRDF, light loop, shadow sampling, IBL, fog, and tonemapping. The Filament documentation makes the contract explicit: "you **must** call `prepareMaterial(material)` before exiting." That single rule is why Filament authoring feels small. The user fills a struct, the engine runs the lighting.

Importantly, **matc is not really a compiler**. The pipeline is: resolve includes inside the material text, run template substitutions keyed by config flags, parse the JSONish metadata into a `MaterialBuilder` AST with full validation, and for each (target, variant) pair hand the assembled GLSL to glslang for real semantic analysis. That's a structured preprocessor with semantic validation, not a new language. The text-template part is the work that anyone honest would call "composition." This is the architecture I want for us.

**Godot's hint annotations** are the gold standard for parameter ergonomics. `uniform vec4 albedo : source_color = vec4(1.0)` gives the engine enough information to do automatic sRGB decode, render an inspector color picker, and provide a sane default. The hint vocabulary is small and worth stealing wholesale: `source_color`, `hint_range(min, max, step)`, `hint_normal`, `hint_default_white`/`hint_default_black`/`hint_default_transparent`, `filter_linear_mipmap`, `repeat_disable`. Even without an inspector, parsed hint metadata gives us a typed Dart constructor with sensible defaults for free.

**Godot's `light()` hook** is the right shape for stylized shading. Omit it and the engine BRDF runs. Provide it and your function replaces the diffuse and specular calculation per-light, while shadows, attenuation, IBL, and the light loop stay engine-owned. A toon shader is a five-line `light()` function, not a from-scratch reimplementation of the lighting pipeline.

**Bevy's `naga_oil`** is the cautionary tale. They compose shaders at the IR level instead of the text level, which gives them dead-code elimination and `virtual fn`/`override fn` across modules. The price is that every editor extension in the WGSL ecosystem squiggles their custom `#import` directives, naga errors reference line numbers in the composed output rather than the user's source, and the canonical community advice for writing a custom material has been "copy the standard PBR shader and edit it" for years. The lesson: keep the surface format close enough to standard GLSL that off-the-shelf editor tooling works. Don't trade ergonomics for clever composition.

**Three.js's `onBeforeCompile`** is the API anti-pattern. It's string-patching of pre-existing shader chunks, order-dependent, undocumented chunk names, and it silently broke on WebGPU. The community has spent seven years migrating off it. Don't expose patching as the customization API; expose explicit hooks.

**Slang** is the most ambitious option, and the right *language* for v4+. Its `interface`/generics system is the genuinely correct abstraction for "engine provides the lighting loop, user writes their BRDF," with specialization at compile time and no preprocessor cleverness. But it doesn't fit Flutter GPU's pipeline today (`impellerc` consumes GLSL or HLSL), so it isn't a starting point.

## What `impellerc` and Flutter GPU already give us

This is the load-bearing fact for the design:

- `impellerc` already accepts `--include=<dir>` (repeatable) and `--define=<MACRO[=value]>` (repeatable). We can host a Filament-style preprocessor by emitting GLSL with appropriate `#define`s and a directory of `#include`-able framework shaders.
- The shader bundle is a flatbuffer with full reflection metadata: uniform names, types, offsets, sizes, descriptor sets, binding slots, sampled-image bindings, stage I/O locations. All preserved through to the Dart side.
- **Flutter GPU resolves uniforms by name at runtime**, not by slot index (`fragmentShader.getUniformSlot("FragInfo")`). v1's `ShaderMaterial` is built entirely on this: it holds `Map<String, ByteData>` uniform blocks and binds by name. A custom-material runtime needs no slot bookkeeping.
- The shader bundle manifest is already a JSON file with `type`, `file`, and optional `language` / `entry_point`. Extending it (or synthesizing one) is straightforward.

What it does **not** give us:

- **Hot reload is landing upstream (this was the single biggest gap).** When this comment was first written, shader iteration meant a full restart. That is changing. Build hooks already rerun on a hot reload tick when a declared input changes, and the Flutter GPU shader hot reload chain is now in flight (umbrella flutter/flutter#186346: engine in-place reload flutter/flutter#186344 via PR flutter/flutter#186793, producer-agnostic `flutter_tools` dispatch flutter/flutter#186345). Because that dispatch is keyed by asset path and no-ops on a registry miss, a flutter_scene material bundle gets hot reload for free once those land and roll into the SDK, with no flutter_scene-side machinery. See v3.
- **No editor.** Flutter has no built-in 3D scene editor, so there's no "right place" to surface uniform hints as UI today. The metadata is worth emitting anyway for the future and for codegen.
- **No specialization constants exposed at the Dart layer** (as far as I can tell). Variant management has to live above `impellerc` for now.

## v1: Foundation (shipped)

Shipped. `ShaderMaterial` takes a `gpu.Shader` from any `flutter_gpu_shaders` bundle plus name-bound parameter values (uniforms resolved by name via `getUniformSlot`, samplers via the existing pattern). The engine BRDF / IBL / tonemap code is shipped as `#include`-able GLSL chunks (`pbr.glsl`, `normals.glsl`, `texture.glsl`, `tone_mapping.glsl`), the per-frame uniform contract is documented in `MATERIALS.md`, and there are worked examples. The user writes raw GLSL and binds parameters by name: full control, manual wiring. This already unblocks the common community cases (toon, rim light, vertex-painted, custom alpha modes). v2 keeps this exact runtime and adds ergonomics on top of it.

## v2: The preprocessor and the `.fs_mat` format

This is where authoring becomes genuinely pleasant. The format is shaped like Filament's `.mat` with Godot's hint annotations grafted on; the preprocessor is a small Dart tool in flutter_scene that the user's build hook invokes in-process.

**Format sketch:**

```glsl
// my_toon.fs_mat
material {
  name: "Toon",
  shadingModel: lit,
  blending: opaque,
  culling: back,
  parameters: [
    { type: vec4,      name: base_color, hint: source_color, default: [1, 1, 1, 1] },
    { type: float,     name: rim_width,  hint: range(0, 1, 0.01), default: 0.2 },
    { type: int,       name: band_count, hint: range(1, 8, 1), default: 3 },
    { type: sampler2d, name: base_color_texture, hint: default_white },
  ],
  requires: [ position, normal, uv0 ],
}

fragment {
  void surface(inout MaterialInputs material) {
    material.baseColor = base_color * texture(base_color_texture, getUV0());
    material.metallic = 0.0;
    material.roughness = 1.0;
    prepareMaterial(material);
  }

  void light(in LightSample sample, inout LightingOutput out) {
    // Optional. Override the per-light BRDF for toon banding.
    float NoL = max(dot(getWorldNormal(), sample.direction), 0.0);
    float banded = floor(NoL * float(band_count)) / float(band_count);
    out.diffuse += banded * sample.color * material.baseColor.rgb;
  }
}
```

**Preprocessor pipeline:**

1. Lex the metadata block into a real AST (parsing with regex would be a category error; see Bevy's naga-oil pain). Validate types, hint compatibility, attribute requirements.
2. Resolve `#include`-equivalents inside the GLSL blocks; preserve `#line` directives at every boundary so `impellerc` errors point at the user's `.fs_mat` by line number. This is the single highest-leverage ergonomics decision in the preprocessor and costs about ten lines of emitter code.
3. Emit two standard GLSL files (vertex + fragment): the parameter UBO declaration plus sampler declarations, the engine framework code (`MaterialInputs`, `prepareMaterial`, input accessors, light loop, BRDF, IBL, tonemap), the user's `surface()` and optional `light()`, and a `main()` that runs everything in order.
4. Hand the emitted GLSL to `impellerc` via a synthesized manifest. The flatbuffer bundle is emitted as usual.

**Integration with `flutter_gpu_shaders`: layer on top, don't fork.** flutter_scene already wraps `buildShaderBundleJson` in its own hook for `base.shaderbundle`. v2 adds a `buildMaterials()` helper that the user calls from their `hook/build.dart` the same way: it preprocesses each `.fs_mat` to GLSL, synthesizes a `.shaderbundle.json` manifest, delegates compilation to `flutter_gpu_shaders`, and declares dependencies on the `BuildOutputBuilder`. flutter_gpu_shaders never learns what `.fs_mat` is, so v2 also inherits its eventual DataAssets and hooks 2.0 migrations (#106, #142) for free.

**Dependency tracking works today, independent of the depfile work.** Because the preprocessor resolves its own `#include`s (step 2, which it must do for `#line` anyway), it knows its entire input set and declares `.fs_mat` + framework chunks + user includes directly to `buildOutput.dependencies`. So correct rebuilds on restart work now and do not wait on the impellerc depfile work (flutter/flutter#186341, bdero/flutter_gpu_shaders#15). That depfile only matters for the v1 raw-`ShaderMaterial` path, where flutter_scene never parses the user's includes.

**Split GLSL emission from Dart codegen into two steps.** These are different build systems. GLSL emission fits the build hook. The typed-Dart-class codegen does not: build hooks emit assets, not pre-compilation Dart source, so codegen needs build_runner or a `dart run flutter_scene:gen` step, and it makes `.fs_mat` a two-reader input. So ship them separately:

- **v2.0**: the format, the preprocessor, and a name-driven `PreprocessedMaterial`. This is trivial because the v1 machinery already binds by name; the user gets the `.fs_mat` authoring win immediately.
- **v2.1**: per-`.fs_mat` typed Dart classes via codegen (autocomplete, type-checked setters). This is the one piece with real cross-build-system friction, so it should not gate v2.0.

**Two hook levels**, mirroring Godot's split: `surface(inout MaterialInputs material)` (required, populate inputs, engine runs the BRDF) and `light(in LightSample, inout LightingOutput)` (optional per-light BRDF override for stylized shading).

**Render-mode tokens** in the metadata block (`blending`, `culling`, `depthWrite`, `depthTest`) get parsed at preprocess time and become the material's runtime configuration. No PSO descriptor authored by the user. Steal Godot's `render_mode` vocabulary.

**Variants:** deliberately resist the variant matrix. Filament's combinatorial grid is something they pay for in binary size and load time. Start with just skinned vs unskinned (we already do this) and push every other dimension into uniform-controlled branches. Modern GPUs handle uniform control flow cheaply.

Rough cost: v2.0 is the preprocessor (lex, parse, validate, emit) plus the framework-shader refactor (splitting `flutter_scene_standard.frag` into composable pieces, which is mostly mechanical since the code already exists). v2.1 codegen is a smaller follow-up.

## v3: Hot reload (now upstream) and polish

The original plan made dev-mode shader hot reload the headline v3 feature and assumed flutter_scene would build a `flutter_scene.devMode()` file-watcher plus a custom "reload from a path that bypasses the asset bundle" entry point in `flutter_gpu_shaders`. **That bespoke work is no longer needed.** The Flutter GPU shader hot reload chain (flutter/flutter#186346) makes reload producer-agnostic at the engine and tool level: edit a shader source, the build hook reruns on the reload tick, the new bundle bytes sync through DevFS, and `flutter_tools` dispatches `ext.ui.gpu.reinitializeShaderLibrary` for the changed `.shaderbundle`, which the engine reloads in place. A v2 material bundle gets this for free; flutter_scene only has to emit a normal bundle at a stable asset path, which it already does, and declare the preprocessor's inputs as dependencies, which v2 already does.

So v3 is just polish:

- Better preprocessor error messages: structured complaints at parse time ("parameter `albedo` has type `vec3` but `source_color` requires `vec4`"), source locations everywhere.
- More worked examples: clearcoat, anisotropic specular, parallax-mapped, vertex-displacement. A "materials gallery" tab in the example app.
- Optional: a small standalone material-preview app (turntable camera) for authoring without a running game.
- Optional: emit `material.json` next to each compiled bundle capturing parameter and hint metadata for future tooling consumers.

## v4+: Future direction

Deferred until v1 through v3 are battle-tested:

- **Slang as the source language.** The right long-term target. The v2 design is forward-compatible: replace the GLSL emitter with a Slang emitter, keep the `.fs_mat` declarative metadata block, keep the runtime mostly unchanged.
- **Visual shader graph.** Worth considering eventually, but the canonical text format needs to land first and stay the authoritative source of truth (Godot's "visual shader to text shader" model is the right pattern).
- **Specialization constants** for runtime variants instead of separate compiled programs. We need to verify what Flutter GPU exposes; if nothing, this becomes an upstream ask.
- **Material inheritance and composition** ("ToonGold extends Toon with metallic=1"). Solvable with codegen on top of v2.

## Open questions worth community input

1. **Parameter accessor syntax.** Filament prefixes user parameters as `materialParams.base_color`. Godot exposes the bare name. I lean toward bare names for ergonomics, but the prefix prevents collisions with engine-provided identifiers.
2. **File extension.** `.fs_mat` is awkward but explicit. `.mat` collides with Filament. `.fsmat`? Naming bikeshed, but the wrong choice is annoying forever.
3. **Should `surface()` be the only required entry point, or also allow a raw `fragment()` escape hatch?** I lean toward both, with raw `fragment()` documented as "you're opting out of the engine lighting pipeline." Same shape as Godot's `unshaded`.
4. **Per-material vs global include paths.** I lean toward relative paths plus an opt-in workspace include root for shared libraries.
5. **Vertex-shader story.** A `.fs_mat` should be able to opt in to a custom vertex function, but the default should be "use the engine's standard vertex shader, no boilerplate."

## Upstream dependencies

The hot reload story rides entirely on work already in flight. None of it blocks v2 from shipping; all of it makes v2 and v3 better as it lands:

- flutter/flutter#186346: umbrella tracker for Flutter GPU shader hot reload.
- flutter/flutter#186344 (PR flutter/flutter#186793): engine in-place `ShaderLibrary` reload, dirty bit, pipeline eviction, and the `ext.ui.gpu.reinitializeShaderLibrary` service extension.
- flutter/flutter#186345: producer-agnostic `flutter_tools` dispatch of that extension for changed `.shaderbundle` assets. This is the piece that lights up hot reload for flutter_scene materials.
- flutter/flutter#186341 (merged): `impellerc` emits a depfile in `--shader-bundle` mode. Relevant to the v1 raw-shader path, not to v2.
- bdero/flutter_gpu_shaders#15: consume that depfile in the build hook. Same scope as above.
- flutter/flutter#187051: let asset transformers declare input dependencies. Only relevant if we ever ship the v2 preprocessor as an asset transformer instead of a build hook.

## Why I think this is the right shape

The recurring failure mode of custom-material APIs is to optimize for engine implementation simplicity at the expense of user friction. Three.js's `ShaderMaterial` did this and the community wrote `THREE-CustomShaderMaterial` to escape. Bevy's naga-oil did this and the community spent a year asking how to write a custom material. Unity's surface shaders did this and the SRP rewrite invalidated everyone's content.

A real preprocessor with a small declarative format costs us a few thousand lines of Dart, but it lets the user write a working, fully-integrated material in twenty lines. It also positions us to migrate the underlying compilation path (to Slang, to a custom IR, to whatever) without breaking the surface format that users author in. The format is the contract; the preprocessor is the implementation. Filament got this right. Godot got this right. We should follow.

Next step: start v2.0 (the `buildMaterials()` helper, the preprocessor, the framework-shader refactor, and the name-driven `PreprocessedMaterial`). v2.1 codegen and the v3 polish follow once the format feels right in practice.

Cross-references:

- [Filament Materials Guide](https://google.github.io/filament/Materials.md.html)
- [Filament `matc` source](https://github.com/google/filament/tree/main/tools/matc)
- [Godot shading language reference](https://docs.godotengine.org/en/stable/tutorials/shaders/shader_reference/shading_language.html)
- [Bevy `naga_oil`](https://github.com/bevyengine/naga_oil)
- [Slang shading language](https://shader-slang.org/)
- [`impellerc` switches](https://github.com/flutter/flutter/blob/main/engine/src/flutter/impeller/compiler/switches.cc)
- [`flutter_gpu_shaders` package](https://pub.dev/packages/flutter_gpu_shaders)

**Comment by bdero** [comment]
We now have fully hot reloadable custom materials that hook into the lighting system.
